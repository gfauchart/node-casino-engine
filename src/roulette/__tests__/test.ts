import RouletteEngine, { SlotBetType } from '../index';

const playerId = '112233';

test('Roulette Engine should throw when player is not registered', () => {
  const engine = new RouletteEngine();
  expect(() =>
    engine.placeBet(playerId, {type : SlotBetType.Straight, value: 2, amount: 2})
  ).toThrowError(/not registered/i);
});

test('Roulette Engine should throw when player does not have enough fund', () => {
  const engine = new RouletteEngine();
  expect(() =>
    engine
      .registerPlayer({playerId, balance: 100 })
      .placeBet(playerId, {type : SlotBetType.Straight, value: 2, amount: 60})
      .placeBet(playerId, {type : SlotBetType.Straight, value: 2, amount: 60})
  ).toThrowError(/do not have enough funds/i);
});

test('Roulette Engine should throw when player is placing a bet on an invalid slot', () => {
  const engine = new RouletteEngine();
  expect(() =>
    engine
      .registerPlayer({playerId, balance: 100 })
      .placeBet(playerId, {type : SlotBetType.Straight, value: 45, amount: 60})
  ).toThrowError(/invalid slot/i);
});

test('Roulette Engine spin should return bets', () => {
  const engine = new RouletteEngine();
  const bets  = engine
    .registerPlayer({playerId, balance: 100 })
    .placeBet(playerId, {type : SlotBetType.Odd, amount: 5})
    .placeBet(playerId, {type : SlotBetType.Even, amount: 5})
    .getBets();

  expect(bets).toHaveLength(2);
  expect(bets).toMatchSnapshot();
});

test('Roulette Engine spin should return bets', () => {
  const engine = new RouletteEngine();
  const bets  = engine
    .registerPlayer({playerId, balance: 100 })
    .placeBet(playerId, {type : SlotBetType.Red, amount: 5})
    .placeBet(playerId, {type : SlotBetType.Black, amount: 5})
    .getBets();

  expect(bets).toHaveLength(2);
  expect(bets).toMatchSnapshot();
});

test('Roulette Engine spin should return only 1 win when betting on odd, even and 0', () => {
  const engine = new RouletteEngine();
  const { bets } = engine
    .registerPlayer({playerId, balance: 100 })
    .placeBet(playerId, {type : SlotBetType.Straight, value: 0, amount: 5})
    .placeBet(playerId, {type : SlotBetType.Odd, amount: 5})
    .placeBet(playerId, {type : SlotBetType.Even, amount: 5})
    .spin();
  const winningBets = bets.filter(b => b.won)
  const lostBets = bets.filter(b => !b.won)
  expect(winningBets).toHaveLength(1);
  expect(lostBets).toHaveLength(2);
});

test('Roulette Engine spin should return only 1 win when betting on red, black and 0', () => {
  const engine = new RouletteEngine();
  const { bets } = engine
    .registerPlayer({playerId, balance: 100 })
    .placeBet(playerId, {type : SlotBetType.Straight, value: 0, amount: 5})
    .placeBet(playerId, {type : SlotBetType.Red, amount: 5})
    .placeBet(playerId, {type : SlotBetType.Black, amount: 5})
    .spin();
  const winningBets = bets.filter(b => b.won)
  const lostBets = bets.filter(b => !b.won)
  expect(winningBets).toHaveLength(1);
  expect(lostBets).toHaveLength(2);
});
