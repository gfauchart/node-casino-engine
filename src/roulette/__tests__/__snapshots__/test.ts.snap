// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Roulette Engine spin should return bets 1`] = `
Array [
  Object {
    "amount": 5,
    "multiplier": 1,
    "playerId": "112233",
    "winningSlots": Array [
      Object {
        "color": 1,
        "number": 2,
      },
      Object {
        "color": 1,
        "number": 4,
      },
      Object {
        "color": 1,
        "number": 6,
      },
      Object {
        "color": 1,
        "number": 8,
      },
      Object {
        "color": 1,
        "number": 10,
      },
      Object {
        "color": 0,
        "number": 12,
      },
      Object {
        "color": 0,
        "number": 14,
      },
      Object {
        "color": 0,
        "number": 16,
      },
      Object {
        "color": 0,
        "number": 18,
      },
      Object {
        "color": 1,
        "number": 20,
      },
      Object {
        "color": 1,
        "number": 22,
      },
      Object {
        "color": 1,
        "number": 24,
      },
      Object {
        "color": 1,
        "number": 26,
      },
      Object {
        "color": 1,
        "number": 28,
      },
      Object {
        "color": 0,
        "number": 30,
      },
      Object {
        "color": 0,
        "number": 32,
      },
      Object {
        "color": 0,
        "number": 34,
      },
      Object {
        "color": 0,
        "number": 36,
      },
    ],
  },
  Object {
    "amount": 5,
    "multiplier": 1,
    "playerId": "112233",
    "winningSlots": Array [
      Object {
        "color": 0,
        "number": 1,
      },
      Object {
        "color": 0,
        "number": 3,
      },
      Object {
        "color": 0,
        "number": 5,
      },
      Object {
        "color": 0,
        "number": 7,
      },
      Object {
        "color": 0,
        "number": 9,
      },
      Object {
        "color": 1,
        "number": 11,
      },
      Object {
        "color": 1,
        "number": 13,
      },
      Object {
        "color": 1,
        "number": 15,
      },
      Object {
        "color": 1,
        "number": 17,
      },
      Object {
        "color": 0,
        "number": 19,
      },
      Object {
        "color": 0,
        "number": 21,
      },
      Object {
        "color": 0,
        "number": 23,
      },
      Object {
        "color": 0,
        "number": 25,
      },
      Object {
        "color": 0,
        "number": 27,
      },
      Object {
        "color": 1,
        "number": 29,
      },
      Object {
        "color": 1,
        "number": 31,
      },
      Object {
        "color": 1,
        "number": 33,
      },
      Object {
        "color": 1,
        "number": 35,
      },
    ],
  },
]
`;

exports[`Roulette Engine spin should return bets 2`] = `
Array [
  Object {
    "amount": 5,
    "multiplier": 1,
    "playerId": "112233",
    "winningSlots": Array [
      Object {
        "color": 0,
        "number": 1,
      },
      Object {
        "color": 0,
        "number": 3,
      },
      Object {
        "color": 0,
        "number": 5,
      },
      Object {
        "color": 0,
        "number": 7,
      },
      Object {
        "color": 0,
        "number": 9,
      },
      Object {
        "color": 0,
        "number": 12,
      },
      Object {
        "color": 0,
        "number": 14,
      },
      Object {
        "color": 0,
        "number": 16,
      },
      Object {
        "color": 0,
        "number": 18,
      },
      Object {
        "color": 0,
        "number": 19,
      },
      Object {
        "color": 0,
        "number": 21,
      },
      Object {
        "color": 0,
        "number": 23,
      },
      Object {
        "color": 0,
        "number": 25,
      },
      Object {
        "color": 0,
        "number": 27,
      },
      Object {
        "color": 0,
        "number": 30,
      },
      Object {
        "color": 0,
        "number": 32,
      },
      Object {
        "color": 0,
        "number": 34,
      },
      Object {
        "color": 0,
        "number": 36,
      },
    ],
  },
  Object {
    "amount": 5,
    "multiplier": 1,
    "playerId": "112233",
    "winningSlots": Array [
      Object {
        "color": 1,
        "number": 2,
      },
      Object {
        "color": 1,
        "number": 4,
      },
      Object {
        "color": 1,
        "number": 6,
      },
      Object {
        "color": 1,
        "number": 8,
      },
      Object {
        "color": 1,
        "number": 10,
      },
      Object {
        "color": 1,
        "number": 11,
      },
      Object {
        "color": 1,
        "number": 13,
      },
      Object {
        "color": 1,
        "number": 15,
      },
      Object {
        "color": 1,
        "number": 17,
      },
      Object {
        "color": 1,
        "number": 20,
      },
      Object {
        "color": 1,
        "number": 22,
      },
      Object {
        "color": 1,
        "number": 24,
      },
      Object {
        "color": 1,
        "number": 26,
      },
      Object {
        "color": 1,
        "number": 28,
      },
      Object {
        "color": 1,
        "number": 29,
      },
      Object {
        "color": 1,
        "number": 31,
      },
      Object {
        "color": 1,
        "number": 33,
      },
      Object {
        "color": 1,
        "number": 35,
      },
    ],
  },
]
`;
